<!DOCTYPE html><html><div class="item-title">
        Item 263
      </div> <div class="item-details"><div><b>git_comments:</b> <ol></ol></div><div><b>git_commits:</b> <ol><li><div><div><b>summary:</b> [CXF-4181] Fix fault qname namespace handling
                </div><div><b>message:</b> [CXF-4181] Fix fault qname namespace handling

git-svn-id: https://svn.apache.org/repos/asf/cxf/trunk@1304535 13f79535-47bb-0310-9956-ffa450edef68

                </div></div></li></ol></div><div><b>github_issues:</b> <ol></ol></div><div><b>github_issues_comments:</b> <ol></ol></div><div><b>github_pulls:</b> <ol></ol></div><div><b>github_pulls_comments:</b> <ol></ol></div><div><b>github_pulls_reviews:</b> <ol></ol></div><div><b>jira_issues:</b> <ol><li><div><div><b>summary:</b> CXF error when parsing a SOAP 1.2 fault: Invalid QName in mapping
                </div><div><b>description:</b> When receiving the following SOAP 1.2 fault, a parsing error occurs in Soap12FaultInInterceptor:
&lt;S:Envelope xmlns:S="http://www.w3.org/2003/05/soap-envelope" 
	xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" 
	xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" 
	xmlns:wst="http://schemas.xmlsoap.org/ws/2005/02/trust" 
	xmlns:psf="http://schemas.microsoft.com/Passport/SoapServices/SOAPFault"&gt;
  &lt;S:Body&gt;
    &lt;S:Fault&gt;
      &lt;S:Code&gt;
	&lt;S:Value&gt;S:Sender&lt;/S:Value&gt;
	&lt;S:Subcode&gt;
	  &lt;S:Value&gt;wst:FailedAuthentication&lt;/S:Value&gt;
	&lt;/S:Subcode&gt;
      &lt;/S:Code&gt;
      &lt;S:Reason&gt;
	&lt;S:Text xml:lang="en-US"&gt;Authentication Failure&lt;/S:Text&gt;
      &lt;/S:Reason&gt;
      &lt;S:Detail&gt;
	&lt;psf:error&gt;
	  &lt;psf:value&gt;0x80048821&lt;/psf:value&gt;
	  &lt;psf:internalerror&gt;
	    &lt;psf:code&gt;0x80041012&lt;/psf:code&gt;
	    &lt;psf:text&gt;The entered and stored passwords do not match.&amp;#x000D;&amp;#x000A;&lt;/psf:text&gt;
	  &lt;/psf:internalerror&gt;
	&lt;/psf:error&gt;
      &lt;/S:Detail&gt;
    &lt;/S:Fault&gt;
  &lt;/S:Body&gt;
&lt;/S:Envelope&gt;

Stack trace:
java.lang.RuntimeException: Invalid QName in mapping: wst:FailedAuthentication
	at org.apache.cxf.helpers.XMLUtils.getQName(XMLUtils.java:447)
	at org.apache.cxf.binding.soap.interceptor.Soap12FaultInInterceptor.unmarshalFault(Soap12FaultInInterceptor.java:88)
	at org.apache.cxf.binding.soap.interceptor.Soap12FaultInInterceptor.handleMessage(Soap12FaultInInterceptor.java:59)
	at org.apache.cxf.binding.soap.interceptor.Soap12FaultInInterceptor.handleMessage(Soap12FaultInInterceptor.java:46)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:263)
	at org.apache.cxf.interceptor.AbstractFaultChainInitiatorObserver.onMessage(AbstractFaultChainInitiatorObserver.java:113)
	at org.apache.cxf.binding.soap.interceptor.CheckFaultInterceptor.handleMessage(CheckFaultInterceptor.java:69)
	at org.apache.cxf.binding.soap.interceptor.CheckFaultInterceptor.handleMessage(CheckFaultInterceptor.java:34)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:263)
	at org.apache.cxf.endpoint.ClientImpl.onMessage(ClientImpl.java:799)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleResponseInternal(HTTPConduit.java:1627)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleResponse(HTTPConduit.java:1494)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.close(HTTPConduit.java:1402)
	at org.apache.cxf.io.CacheAndWriteOutputStream.postClose(CacheAndWriteOutputStream.java:47)
	at org.apache.cxf.io.CachedOutputStream.close(CachedOutputStream.java:195)
	at org.apache.cxf.transport.AbstractConduit.close(AbstractConduit.java:56)
	at org.apache.cxf.transport.http.HTTPConduit.close(HTTPConduit.java:649)
	at org.apache.cxf.interceptor.MessageSenderInterceptor$MessageSenderEndingInterceptor.handleMessage(MessageSenderInterceptor.java:62)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:263)
	at org.apache.cxf.endpoint.ClientImpl.doInvoke(ClientImpl.java:533)
	at org.apache.cxf.endpoint.ClientImpl.invoke(ClientImpl.java:463)
	at org.apache.cxf.endpoint.ClientImpl.invoke(ClientImpl.java:366)
	at org.apache.cxf.endpoint.ClientImpl.invoke(ClientImpl.java:319)
	at org.apache.cxf.ws.security.trust.STSClient.requestSecurityToken(STSClient.java:708)
	at org.apache.cxf.ws.security.trust.STSClient.requestSecurityToken(STSClient.java:584)
	at org.apache.cxf.ws.security.trust.STSClient.requestSecurityToken(STSClient.java:576)
	at net.entropysoft.eci.sharepoint.webservices.SharepointWebServices.authenticateToSharePointOnline(SharepointWebServices.java:426)

This happens because Soap12FaultInInterceptor creates a new DOM document with the Fault element but omits the namespace declarations on parent nodes (Envelope and Body).


                </div></div></li></ol></div><div><b>jira_issues_comments:</b> <ol><li><div>
                This needs some change in the way how the soap fault is built.
I thought we can just get all the bindings from reader.getNamespaceContext() to the root element. But there seems to be no method to get all the bindings.

I am attaching a unit test to demonstrate this problem.

              </div></li><li><div><div><b>body:</b> a unit test annotated with @Ignore for now.
                </div><div><b>label:</b> test
                </div></div></li><li><div>
                Fixing test
              </div></li><li><div>
                Hello I'm a CXF user. I'm encountering the same issue, from a 2.7.3 CXF client (same stacktrace). The Subcode Value namespace prefix is declared on the envelope. This SoapFault is W3C compliant (https://www.w3.org/TR/soap12-part1/#faultcodes). How can we get this soapFault marshalling work ?
              </div></li></ol></div></div></html>